dvc/repo/index.py

```diff
@@ -0,0 +1,275 @@
+from functools import partialmethod
+from typing import (
+    TYPE_CHECKING,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+)
+
+from funcy import cached_property
+
+if TYPE_CHECKING:
+    from networkx import DiGraph
+    from pygtrie import Trie
+    from pyrsistent import PSet
+
+    from dvc.dependency import ParamsDependency
+    from dvc.fs.base import BaseFileSystem
+    from dvc.output import ObjectCollection, Output
+    from dvc.repo.stage import StageLoad
+    from dvc.stage import Stage
+    from dvc.types import StrPath, TargetType
+
+
+class Index:
+    def __init__(
+        self,
+        repo: "Repo",  # pylint: disable=redefined-outer-name
+        fs: "BaseFileSystem",
+        stages: "PSet['Stage']" = None,
+    ) -> None:
+        """Index is an immutable collection of stages.
+
+        Generally, Index is a complete collection of stages at a point in time.
+        With "a point in time", it means it is collected from the user's
+        workspace or a git revision.
+        And, since Index is immutable, the collection is frozen in time.
+
+        Index provides multiple ways to view this collection:
+
+            stages - provides direct access to this collection
+            outputs - provides direct access to the outputs
+            objects - provides direct access to the objects
+            graph -
+            ... and many more.
+
+        Index also provides ways to slice and dice this collection.
+        Some `views` might not make sense when sliced (eg: pipelines/graph).
+        """
+
+        self.fs: "BaseFileSystem" = fs
+        self.repo: "Repo" = repo
+        self.stage_collector: "StageLoad" = repo.stage
+        if stages is not None:
+            self._stages: "PSet['Stage']" = stages
+
+    @cached_property
+    def _stages(self) -> "PSet['Stage']":
+        """Index without stages cannot exist.
+        So, stages being lazy, should not be depended upon.
+        """
+        from pyrsistent import pset
+
+        onerror = self.repo.stage_collection_error_handler
+        # pylint: disable=protected-access
+        return pset(self.stage_collector._collect_repo(onerror=onerror))
+
+    @property
+    def stages(self) -> "PSet['Stage']":
+        return self._stages
```

Comment:
Why do we need `._stages` prop? Why don't move everything here and make it cached property?